File: playground/BOJ/boj_2738/boj_2738.js
/* ----------------------------------------------
Sub: [BOJ] 2738. 행렬 덧셈
Link: https://www.acmicpc.net/problem/2738
Tag: js, 2차원 배열
Memo
---------------------------------------------- */

const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const [N, M] = input[0].split(" ").map(Number);

// 행렬 A와 B 가져오기
let A = [];
let B = [];

for (let i = 1; i <= N; i++) {
  A.push(input[i].split(" ").map(Number));
}

for (let i = N + 1; i <= 2 * N; i++) {
  B.push(input[i].split(" ").map(Number));
}

// 결과 행렬 계산
let result = Array.from(Array(N), () => new Array(M));

for (let i = 0; i < N; i++) {
  for (let j = 0; j < M; j++) {
    result[i][j] = A[i][j] + B[i][j];
  }
}

// 결과 출력
result.forEach((row) => console.log(row.join(" ")));

File: boj/boj_2738/boj_2738.c
/* ----------------------------------------------
Sub: [BOJ] 2738. 행렬 덧셈
Link: https://www.acmicpc.net/problem/2738
Tag: C, 구현, 동적 2차배열
Memo
---------------------------------------------- */

#include <stdio.h>
#include <stdlib.h>

// 동적 2차원 배열 생성
int **create2DArray(int rows, int cols) {
  int **array = (int **)malloc(rows * sizeof(int *));

  for (int i = 0; i < rows; i++) {
    array[i] = (int *)malloc(cols * sizeof(int));
  }

  return array;
}

// 메모리 해제
void free2DArray(int **array, int rows) {
  for (int i = 0; i < rows; i++) {
    free(array[i]);
  }
  free(array);
}

int main() {
  int n, m;
  scanf("%d %d", &n, &m);

  int **arrayA = create2DArray(n, m);
  int **arrayB = create2DArray(n, m);

  // 행렬A 입력
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < m; j++) {
      scanf("%d", &arrayA[i][j]);
    }
  }

  // 행렬B 입력
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < m; j++) {
      scanf("%d", &arrayB[i][j]);
    }
  }

  // 행렬A + 행렬B 결과값 출력
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < m; j++) {
      printf("%d ", arrayA[i][j] + arrayB[i][j]);
    }
    printf("\n");
  }

  free2DArray(arrayA, n);
  free2DArray(arrayB, n);

  return 0;
}
